### PR TITLE
Add C# feature "Expression body syntax for property accessors"

### LINKED ISSUE #7881: Proposal: Expression bodied get and set methods
C# 6.0 allowed get-only properties and all methods to have an expression body. This is a great feature that can help keep the focus on business logic rather than code. However, when you need to fall back to writing a `get` and `set` method manually, it can be annoying that you're no longer allowed to use expression bodies.

I propose supporting expression bodies for `get` and `set` methods. It makes code less verbose and also would provide more symmetry in the language.

``` C#
class Foo
{
    int Bar
    {
        get => _bar;
        set => _bar = value;
    }
    private int _bar;
}
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #7881

@dotnet/roslyn-compiler Please review

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs [hunk 1]
```diff
@@ -192,12 +194,14 @@ internal override bool IsExpressionBodied
...
             DiagnosticBag diagnostics) :
-            base(containingType, syntax.GetReference(), syntax.Body?.GetReference(), location)
+            base(containingType, syntax.GetReference(), syntax.Body?.GetReference() ?? syntax.ExpressionBody?.GetReference(), location)
         {
             _property = property;
             _explicitInterfaceImplementations = explicitInterfaceImplementations;
             _name = name;
             _isAutoPropertyAccessor = isAutoPropertyAccessor;
+            Debug.Assert(_property.IsExpressionBodied == false, "Cannot have accessors in expression bodied lightweight properties");
+            _isExpressionBodied = (syntax.Body == null && syntax.ExpressionBody != null);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: >syntax.Body == null [](start = 35, length = 19)

Consider factoring `syntax.Body == null`. Then you don't need the `bodyOpt` local below.

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: >syntax.ExpressionBody [](start = 58, length = 21)

Also, `syntax.ExpressionBody` is `expressionBodyOpt` below.

**Comment by jcouv** [comment]
Code looks good, but need to add tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ExpressionBodiedMemberTests.cs [hunk 2]
```diff
@@ -481,28 +489,37 @@ public void BlockBodyAndExpressionBody_04()
...
-    // (8,10): error CS1519: Invalid token ';' in class, struct, or interface member declaration
-    //     => P1;
-    Diagnostic(ErrorCode.ERR_InvalidMemberDecl, ";").WithArguments(";").WithLocation(8, 10),
-    // (8,10): error CS1519: Invalid token ';' in class, struct, or interface member declaration
-    //     => P1;
-    Diagnostic(ErrorCode.ERR_InvalidMemberDecl, ";").WithArguments(";").WithLocation(8, 10)
+    // (5,5): error  CS8057: Methods cannot combine block bodies with expression bodies.
+    Diagnostic(ErrorCode.ERR_BlockBodyAndExpressionBody, @"~C()
+    { P1 = 1; }
+    => P1;").WithLocation(6, 5)
```
Comment (hughbe) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ExpressionBodiedMemberTests.cs]: Minor: Comment above "(5,5)" doesn't seem to line up with the `WithLocation(6,5)`

**Comment by AlekseyTs** [comment]
Unless I am missing something, there are no new tests. I would expect to see a lot of new tests, including the ones targeting SemanticModel.

**Comment by gafter** [comment]
I've filed #13578 tracking the needed tests so we can get this code into the feature branch.

---
In reply to: [244451551](https://github.com/dotnet/roslyn/pull/13543#issuecomment-244451551) [](ancestors = 244451551)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertyAccessorSymbol.cs]: I'll refactor that in the next PR (e.g. when adding tests)

---
In reply to: [77268250](https://github.com/dotnet/roslyn/pull/13543#discussion_r77268250) [](ancestors = 77268250)

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews for placing this into its feature branch?

**Comment by gafter** [comment]
The implementation LGTM for integration into the feature branch. We need tests (#13578) to go from the feature branch into master.

@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review of the feature implementation for integration into its feature branch? Issue #13578 is tracking the needed tests so we can get this increment of the implementation into the feature branch and complete the feature in follow-up PRs.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -210,6 +210,12 @@
   <data name="IDS_FeaturePatternMatching" xml:space="preserve">
     <value>pattern matching</value>
   </data>
+  <data name="IDS_FeatureExpressionBodiedAccessor" xml:space="preserve">
+    <value>expression body property accessor</value>
+  </data>
+  <data name="IDS_FeatureExpressionBodiedDeOrConstructor" xml:space="preserve">
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Typo: `...DeOrConstructor`

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 4]
```diff
@@ -2416,7 +2416,7 @@ static void M()
 
         // TODO: extra error CS1014
         [Fact]
-        public void CS1043ERR_SemiOrLBraceExpected()
+        public void CS7887ERR_SemiOrLBraceOrArrowExpected()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: Do we still have a test for `CS1043`?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: That's what it says. That is shorthand for Deconstructor Or Constructor.

---
In reply to: [77881977](https://github.com/dotnet/roslyn/pull/13543#discussion_r77881977) [](ancestors = 77881977)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: Yes, but we can probably retire the old code and uniformly use the new error code.

---
In reply to: [77885635](https://github.com/dotnet/roslyn/pull/13543#discussion_r77885635) [](ancestors = 77885635)

**Pull request merged**